### PR TITLE
feat(account invitation): implement sign in before accepting the invitation

### PR DESCRIPTION
**Pull request by andreybakanovsky**

### CODE REVIEW

File: app/controllers/accept_account_invitations_controller.rb [hunk 1]
```diff
@@ -1,18 +1,24 @@
...
+    return if user_signed_in? && the_invitation
 
-  def show; end
+    session[:token] = ps[:token]
+    redirect_to new_user_session_url(email: the_invitation.email)
+  end
 
   def update
     received_invitation.update!(accepted_at: Time.current)
-    redirect_to account_path(received_invitation.account_id)
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:10]: I agree that it still was not working, but it was the correct line. Revert it back? Why removed?

File: app/controllers/accept_account_invitations_controller.rb [hunk 2]
```diff
@@ -1,18 +1,24 @@
 # frozen_string_literal: true
 
 class AcceptAccountInvitationsController < ApplicationController
-  before_action :authenticate_user!
+  def show
+    return if user_signed_in? && the_invitation
 
-  def show; end
+    session[:token] = ps[:token]
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:7]: Better remember `after_sign_in_url` in the session, this way the after-sign-in redirect implementation will be easy without knowing the internals of possible workflows current or future.

File: app/controllers/accept_account_invitations_controller.rb [hunk 3]
```diff
@@ -1,18 +1,24 @@
...
-    redirect_to account_path(received_invitation.account_id)
   end
 
   private
 
   helper_method memoize def received_invitation
     AccountInvitation.unaccepted.for(current_user).find_by!(token: ps.fetch(:token))
   end
+
+  memoize def the_invitation
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:21]: Remove this check, if they sign in and the after-sign-in link is wrong, they will see 404 page and this is fine. In other words, the code already handles this scenario and no need to write special code for that.

File: app/controllers/users/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  class SessionsController < Devise::SessionsController
+    def new
+      if params.key?(:email)
+        self.resource = User.new(email: params[:email])
+      else
+        super
+      end
+    end
+
+    protected
```
Comment (ka8725) [inline app/controllers/users/sessions_controller.rb:13]:
```suggestion
    private
```

Learn about `protected` in Ruby. It's not the same as in the common OOP (like Java).

**Review by ka8725 — CHANGES_REQUESTED** [review]

File: app/controllers/users/sessions_controller.rb (hunk 4 above)
Comment (andreybakanovsky) [reply app/controllers/users/sessions_controller.rb:13]: I'm aware about. It's just missprint

File: app/controllers/accept_account_invitations_controller.rb (hunk 1 above)
Comment (andreybakanovsky) [reply app/controllers/accept_account_invitations_controller.rb]: Ok, let's keep

File: app/controllers/accept_account_invitations_controller.rb (hunk 2 above)
Comment (andreybakanovsky) [reply app/controllers/accept_account_invitations_controller.rb]: Global. Beautiful.

**Review by VladislavSokov — APPROVED** [review]

File: app/controllers/users/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        super
+      end
+    end
+
+    private
+
+    def after_sign_in_path_for(_resource)
+      if (after_sign_in_url = session.delete(:after_sign_in_url)).present?
+        return after_sign_in_url
+      end
```
Comment (ka8725) [inline app/controllers/users/sessions_controller.rb:18]:
```suggestion
      after_sign_in_url = session.delete(:after_sign_in_url)
      return after_sign_in_url if after_sign_in_url.present?
```

(don't assign have assignment inside boolean expressions, these side-effects are tricky to read and debug)

**Pull request merged**